### PR TITLE
MGMT-16624: Refactor post pivot systemd services

### PR DESCRIPTION
**Pull request by tsorya**
[MGMT-16624](https://issues.redhat.com//browse/MGMT-16624): Refactor post pivot systemd services
1. Removing prepare-installation script
2. Bash code moved to golang in post-pivot

### CODE REVIEW

File: lca-cli/postpivot/postpivot.go [hunk 1]
```diff
@@ -175,6 +175,18 @@ func (p *PostPivot) recert(ctx context.Context, seedReconfiguration *clusterconf
...
+	_ = wait.PollUntilContextCancel(ctxWithTimeout, time.Second, true, func(ctx context.Context) (bool, error) {
+		p.log.Info("pulling recert image")
+		if _, err := p.ops.RunInHostNamespace("podman", "pull", seedClusterInfo.RecertImagePullSpec); err != nil {
+			p.log.Warnf("failed to pull recert image, will retry, err: %s", err.Error())
+			return false, nil
+		}
+		// TODO: add waiting for block device with label
+		return true, nil
+	})
+	defer cancel()
```
Comment (leo8a) [inline lca-cli/postpivot/postpivot.go:188]: wondering if `cancel()` should be called before the `wait`? 

...just to ensure that the cancellation function is invoked even if an error occurs before the polling loop starts

Comment (tsorya) [reply lca-cli/postpivot/postpivot.go:188]: defer should be added before it for sure, will change

File: lca-cli/postpivot/postpivot.go [hunk 2]
```diff
@@ -602,3 +604,141 @@ func (p *PostPivot) createPullSecretManifest(pullSecret, pullSecretManifest stri
 	}
 	return nil
 }
+
+// waifForConfiguration waits till configuration folder exists or till device with label "relocation-config" is added.
+// in case device was provided we are mounting it and copying files to configFolder
+func (p *PostPivot) waifForConfiguration(ctx context.Context, configFolder string) error {
+	if _, err := os.Stat(configFolder); err == nil {
```
Comment (eranco74) [inline lca-cli/postpivot/postpivot.go:611]: you have it in line 618, why do we need it here?

File: lca-cli/postpivot/postpivot.go [hunk 3]
```diff
@@ -602,3 +604,141 @@ func (p *PostPivot) createPullSecretManifest(pullSecret, pullSecretManifest stri
...
+		}
+		for _, bd := range blockDevices {
+			if bd.Label == label {
+				if err := p.setupConfigurationFolder(bd.Name, "/mnt/config", filepath.Dir(configFolder)); err != nil {
+					return false, err
+				}
+				return true, nil
+			}
+		}
+		// TODO: add waiting for block device with label
```
Comment (eranco74) [inline lca-cli/postpivot/postpivot.go:634]: What's this for?

File: lca-cli/postpivot/postpivot.go [hunk 4]
```diff
@@ -602,3 +604,141 @@ func (p *PostPivot) createPullSecretManifest(pullSecret, pullSecretManifest stri
...
+			return true, nil
+		}
+		blockDevices, err := p.ops.ListBlockDevices()
+		if err != nil {
+			p.log.Infof("Failed to list block devices with error %s, will retry", err.Error())
+			return false, nil
+		}
+		for _, bd := range blockDevices {
+			if bd.Label == label {
+				if err := p.setupConfigurationFolder(bd.Name, "/mnt/config", filepath.Dir(configFolder)); err != nil {
```
Comment (eranco74) [inline lca-cli/postpivot/postpivot.go:628]: log the error, otherwise, it will be ignored (you return false)

File: lca-cli/postpivot/postpivot.go [hunk 5]
```diff
@@ -602,3 +604,141 @@ func (p *PostPivot) createPullSecretManifest(pullSecret, pullSecretManifest stri
...
+
+	if err := p.ops.Mount(deviceName, mountFolder); err != nil {
+		return err
+	}
+
+	if err := utils.CopyFileIfExists(mountFolder, configFolder); err != nil {
+		return fmt.Errorf("failed to copy contert of %s to %s, err: %w", mountFolder, configFolder, err)
+	}
+
+	defer func() {
```
Comment (eranco74) [inline lca-cli/postpivot/postpivot.go:656]: add defer after each step remove all after the MkdirAll and Umount after Mount

File: lca-cli/postpivot/postpivot.go [hunk 6]
```diff
@@ -175,6 +175,18 @@ func (p *PostPivot) recert(ctx context.Context, seedReconfiguration *clusterconf
...
 	}
 
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, 10*time.Minute)
+	_ = wait.PollUntilContextCancel(ctxWithTimeout, time.Second, true, func(ctx context.Context) (bool, error) {
+		p.log.Info("pulling recert image")
+		if _, err := p.ops.RunInHostNamespace("podman", "pull", seedClusterInfo.RecertImagePullSpec); err != nil {
+			p.log.Warnf("failed to pull recert image, will retry, err: %s", err.Error())
+			return false, nil
+		}
+		// TODO: add waiting for block device with label
```
Comment (tsorya) [inline lca-cli/postpivot/postpivot.go:185]: delete

File: lca-cli/postpivot/postpivot.go [hunk 7]
```diff
@@ -602,3 +604,141 @@ func (p *PostPivot) createPullSecretManifest(pullSecret, pullSecretManifest stri
...
+		// TODO: should we fail if umount fails?
+		_ = p.ops.Umount(deviceName)
+		_ = os.RemoveAll(mountFolder)
+	}()
+
+	return nil
+}
+
+// cleanupNMConnections removes old *.nmconnection files
+func (p *PostPivot) cleanupNMConnections(nmconnectionsFolder string) error {
```
Comment (eranco74) [inline lca-cli/postpivot/postpivot.go:666]: This should happen when we create the seed image not here

File: lca-cli/postpivot/postpivot.go (hunk 3 above)
Comment (tsorya) [inline lca-cli/postpivot/postpivot.go:634]: delete

File: lca-cli/postpivot/postpivot.go [hunk 8]
```diff
@@ -602,3 +604,141 @@ func (p *PostPivot) createPullSecretManifest(pullSecret, pullSecretManifest stri
...
+	p.log.Infof("Removing seed nmconnection files")
+	files, err := filepath.Glob(filepath.Join(nmconnectionsFolder, "*.nmconnection"))
+	if err != nil {
+		return fmt.Errorf("failed to find nmconnection files in %s, %w", nmconnectionsFolder, err)
```
Comment (eranco74) [inline lca-cli/postpivot/postpivot.go:670]: Why this is an error?

File: internal/clusterconfig/clusterconfig.go [hunk 9]
```diff
@@ -368,7 +366,7 @@ func (r *UpgradeClusterConfigGather) fetchCABundle(ctx context.Context, manifest
 // gather network files and copy them
 func (r *UpgradeClusterConfigGather) fetchNetworkConfig(ostreeDir string) error {
 	r.Log.Info("Fetching node network files")
-	dir := filepath.Join(ostreeDir, networkDir)
+	dir := filepath.Join(ostreeDir, filepath.Join(common.OptOpenshift, common.NetworkDir))
```
[the remark this answers is not in the document]
Comment (tsorya) [reply internal/clusterconfig/clusterconfig.go:369]: In IBU process we just copy nmconnection files to /opt/openshift/network... so currently we expect them to be there. I didn't add new logic just took the one from script.

File: lca-cli/postpivot/postpivot.go [hunk 10]
```diff
@@ -591,3 +601,112 @@ func (p *PostPivot) createPullSecretManifest(pullSecret, pullSecretManifest stri
 	}
 	return nil
 }
+
+// waifForConfiguration waits till configuration folder exists or till device with label "relocation-config" is added.
+// in case device was provided we are mounting it and copying files to configFolder
+func (p *PostPivot) waifForConfiguration(ctx context.Context, configFolder, blockDeviceMountFolder string) error {
+	err := wait.PollUntilContextCancel(ctx, time.Second, true, func(ctx context.Context) (bool, error) {
```
Comment (eranco74) [inline lca-cli/postpivot/postpivot.go:608]: What happens in case of error?

Comment (eranco74) [reply lca-cli/postpivot/postpivot.go:608]: Well, mount may fail, why not retry?

Comment (tsorya) [reply lca-cli/postpivot/postpivot.go:608]: We can, i thought that it sort of error that you will just always fail no?

**Review by eranco74 — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**